src/puppetlabs/puppetdb/query_eng.clj

```diff
@@ -84,3 +88,39 @@
         (do (log/debug e "Caught PSQL processing exception")
             (http/error-response (.getMessage e)))
         (throw e)))))
+
+(pls/defn-validated object-exists? :- s/Bool
+  "Returns true if an object exists."
+  [type :- s/Keyword
+   id :- s/Str]
+  (let [check-sql (case type
+                    :catalog "SELECT 1
+                              FROM certnames
+                              INNER JOIN catalogs
+                                ON catalogs.certname = certnames.certname
+                              WHERE certnames.certname=?
+                                AND deactivated IS NULL
+                                AND expired IS NULL LIMIT 1"
+                    :node "SELECT 1
+                           FROM certnames
+                           WHERE certname=?
+                             AND deactivated IS NULL
+                             AND expired IS NULL LIMIT 1"
+                    :report (str "SELECT 1
+                                  FROM reports
+                                  WHERE " (su/sql-hash-as-str "hash") "=?
+                                  LIMIT 1")
+                    :environment "SELECT 1
+                                  FROM environments
+                                  WHERE name=?
+                                  LIMIT 1"
+                    :factset "SELECT 1
+                              FROM certnames
+                              INNER JOIN factsets
+                              ON factsets.certname = certnames.certname
+                              WHERE certnames.certname=?
+                                AND deactivated IS NULL
+                               AND expired IS NULL LIMIT 1")]
```

Comment:
I think our uniqueness constraints should make the limits unneccesary in each of these cases